// ts-jest/dist/transpiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var logger_1 = require("./utils/logger");
function transpileTypescript(filePath, fileSrc, compilerOptions) {
    logger_1.logOnce('Compiling via normal transpileModule call');
    var _a = transpileViaTranspileModule(filePath, fileSrc, compilerOptions), code = _a.outputText, map = _a.sourceMapText;
    return { code: code, map: map };
}
exports.transpileTypescript = transpileTypescript;
function transpileViaTranspileModule(filePath, fileSource, compilerOptions) {
    var transpileOutput = ts.transpileModule(fileSource, {
        compilerOptions: compilerOptions,
        fileName: filePath,
        reportDiagnostics: true,
    });
    var diagnostics = transpileOutput.diagnostics;
    if (diagnostics.length > 0) {
        var errors = formatDiagnostics(diagnostics);
        logger_1.logOnce(logMessageForTranspilationErrors(errors));
        throw createTranspilationError(errors);
    }
    return transpileOutput;
}
function formatDiagnostics(diagnostics) {
    return diagnostics.map(function (d) { return d.messageText; }) + "\n";
}
function createTranspilationError(errors) {
    return Error("TypeScript compiler encountered syntax errors while transpiling. Errors: " + errors);
}
function logMessageForTranspilationErrors(errors) {
    return "Diagnostic errors from TSC: " + errors;
}
